File: asmif.c
/* asmif.c */
/*****************************************************************************/
/* SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only                     */
/*                                                                           */
/* AS-Portierung                                                             */
/*                                                                           */
/* Befehle zur bedingten Assemblierung                                       */
/*                                                                           */
/*****************************************************************************/

#include "stdinc.h"
#include <string.h>
#include <ctype.h>

#include "bpemu.h"
#include "chunks.h"
#include "strutil.h"
#include "errmsg.h"
#include "asmdef.h"
#include "asmsub.h"
#include "asmpars.h"

#include "asmif.h"


PIfSave FirstIfSave;
Boolean IfAsm;       /* FALSE: in einer neg. IF-Sequenz-->kein Code */

static Boolean ActiveIF;

/*!------------------------------------------------------------------------
 * \fn     ifsave_create(tIfState if_state, Boolean case_found)
 * \brief  create new IF stack element
 * \param  if_state initial state of related construct
 * \param  case_found in block enabled by IF/CASE?
 * \return * to new element
 * ------------------------------------------------------------------------ */

static PIfSave ifsave_create(tIfState if_state, Boolean case_found)
{
  PIfSave p_save;

  p_save = (PIfSave) malloc(sizeof(TIfSave));
  if (p_save)
  {
    p_save->Next = NULL;
    p_save->NestLevel = SaveIFs() + 1;
    p_save->SaveIfAsm = IfAsm;
    as_tempres_ini(&p_save->SaveExpr);
    p_save->CaseFound = case_found;
    p_save->State = if_state;
    p_save->StartLine = CurrLine;
  }
  return p_save;
}

/*!------------------------------------------------------------------------
 * \fn     ifsave_free(PIfSave p_save)
 * \brief  destroy IF stack element
 * \param  p_save element to destroy
 * ------------------------------------------------------------------------ */

static void ifsave_free(PIfSave p_save)
{
  as_tempres_free(&p_save->SaveExpr);
  free(p_save);
}

static LongInt GetIfVal(const tStrComp *pCond)
{
  Boolean IfOK;
  tSymbolFlags Flags;
  LongInt Tmp;

  Tmp = EvalStrIntExpressionWithFlags(pCond, Int32, &IfOK, &Flags);
  if (mFirstPassUnknown(Flags) || !IfOK)
  {
    Tmp = 1;
    if (mFirstPassUnknown(Flags)) WrError(ErrNum_FirstPassCalc);
  }

  return Tmp;
}


static void AddBoolFlag(Boolean Flag)
{
  strmaxcpy(ListLine, Flag ? "=>TRUE" : "=>FALSE", STRINGSIZE);
}


static void PushIF(LongInt IfExpr)
{
  PIfSave NewSave = ifsave_create(IfState_IFIF, IfExpr != 0);

  NewSave->Next = FirstIfSave;
  FirstIfSave = NewSave;
  IfAsm = IfAsm && NewSave->CaseFound;
}


static void CodeIF(void)
{
  LongInt IfExpr;

  ActiveIF = IfAsm;

  if (!IfAsm)
    IfExpr = 1;
  else if (!ChkArgCnt(1, 1))
    IfExpr = 1;
  else
    IfExpr = GetIfVal(&ArgStr[1]);
  if (IfAsm)
    AddBoolFlag(IfExpr != 0);
  PushIF(IfExpr);
}


static void CodeIFDEF(Word Negate)
{
  LongInt IfExpr;
  Boolean Defined;

  ActiveIF = IfAsm;

  if (!IfAsm) IfExpr = 1;
  else if (!ChkArgCnt(1, 1))
    IfExpr = 1;
  else
  {
    Defined = IsSymbolDefined(&ArgStr[1]);
    if (IfAsm)
      strmaxcpy(ListLine, (Defined) ? "=>DEFINED" : "=>UNDEFINED", STRINGSIZE);
    if (!Negate)
      IfExpr = (Defined) ? 1 : 0;
    else
      IfExpr = (Defined) ? 0 : 1;
  }
  PushIF(IfExpr);
}


static void CodeIFUSED(Word Negate)
{
  LongInt IfExpr;
  Boolean Used;

  ActiveIF = IfAsm;

  if (!IfAsm)
    IfExpr = 1;
  else if (!ChkArgCnt(1, 1))
    IfExpr = 1;
  else
  {
    Used = IsSymbolUsed(&ArgStr[1]);
    if (IfAsm)
      strmaxcpy(ListLine, (Used) ? "=>USED" : "=>UNUSED", STRINGSIZE);
    if (!Negate)
      IfExpr = (Used) ? 1 : 0;
    else
      IfExpr = (Used) ? 0 : 1;
  }
  PushIF(IfExpr);
}


void CodeIFEXIST(Word Negate)
{
  LongInt IfExpr;
  Boolean Found;
  String NPath;

  ActiveIF = IfAsm;

  if (!IfAsm)
    IfExpr = 1;
  else if (!ChkArgCnt(1, 1))
    IfExpr = 1;
  else
  {
    String FileName, Dummy;

    strmaxcpy(FileName, (ArgStr[1].str.p_str[0] == '"') ? ArgStr[1].str.p_str + 1 : ArgStr[1].str.p_str, STRINGSIZE);
    if (FileName[strlen(FileName) - 1] == '"')
      FileName[strlen(FileName) - 1] = '\0';
    AddSuffix(FileName, IncSuffix);
    strmaxcpy(NPath, IncludeList, STRINGSIZE);
    strmaxprep(NPath, SDIRSEP, STRINGSIZE);
    strmaxprep(NPath, ".", STRINGSIZE);
    Found = !FSearch(Dummy, sizeof(Dummy), FileName, CurrFileName, NPath);
    if (IfAsm)
      strmaxcpy(ListLine, Found ? "=>FOUND" : "=>NOT FOUND", STRINGSIZE);
    IfExpr = Negate ? !Found : Found;
  }
  PushIF(IfExpr);
}


static void CodeIFB(Word Negate)
{
  Boolean Blank = True;
  LongInt IfExpr;
  int z;

  ActiveIF = IfAsm;

  if (!IfAsm)
    IfExpr = 1;
  else
  {
    for (z = 1; z <= ArgCnt; z++)
      if (strlen(ArgStr[z++].str.p_str) > 0)
        Blank = False;
    if (IfAsm)
      strmaxcpy(ListLine, (Blank) ? "=>BLANK" : "=>NOT BLANK", STRINGSIZE);
    IfExpr = Negate ? !Blank : Blank;
  }
  PushIF(IfExpr);
}


static void CodeELSEIF(void)
{
  LongInt IfExpr;

  if (!FirstIfSave || (FirstIfSave->State != IfState_IFIF)) WrError(ErrNum_MissingIf);
  else if (ArgCnt == 0)
  {
    if (FirstIfSave->SaveIfAsm)
      AddBoolFlag(IfAsm = (!FirstIfSave->CaseFound));
    FirstIfSave->State = IfState_IFELSE;
  }
  else if (ArgCnt == 1)
  {
    if (!FirstIfSave->SaveIfAsm)
      IfExpr = 1;
    else if (FirstIfSave->CaseFound)
      IfExpr = 0;
    else
      IfExpr = GetIfVal(&ArgStr[1]);
    IfAsm = ((FirstIfSave->SaveIfAsm) && (IfExpr != 0) && (!FirstIfSave->CaseFound));
    if (FirstIfSave->SaveIfAsm)
      AddBoolFlag(IfExpr != 0);
    if (IfExpr != 0)
      FirstIfSave->CaseFound = True;
  }
  else
    (void)ChkArgCnt(0, 1);

  ActiveIF = (!FirstIfSave) || (FirstIfSave->SaveIfAsm);
}


static void CodeENDIF(void)
{
  if (!ChkArgCnt(0, 0));
  else if (!FirstIfSave || ((FirstIfSave->State != IfState_IFIF) && (FirstIfSave->State != IfState_IFELSE))) WrError(ErrNum_MissingIf);
  else
  {
    PIfSave NewSave;

    IfAsm = FirstIfSave->SaveIfAsm;
    NewSave = FirstIfSave;
    FirstIfSave = NewSave->Next;
    as_snprintf(ListLine, STRINGSIZE, "[%u]", (unsigned)NewSave->StartLine);
    ifsave_free(NewSave);
  }

  ActiveIF = IfAsm;
}


static void EvalIfExpression(const tStrComp *pCond, TempResult *erg)
{
  EvalStrExpression(pCond, erg);
  if ((erg->Typ == TempNone) || mFirstPassUnknown(erg->Flags))
  {
    as_tempres_set_int(erg, 1);
    if (mFirstPassUnknown(erg->Flags))
      WrError(ErrNum_FirstPassCalc);
  }
}


static void CodeSWITCH(void)
{
  PIfSave NewSave = ifsave_create(IfState_CASESWITCH, False);

  ActiveIF = IfAsm;

  if (ArgCnt != 1)
  {
    as_tempres_set_int(&NewSave->SaveExpr, 1);
    if (IfAsm)
      (void)ChkArgCnt(1, 1);
  }
  else
  {
    EvalIfExpression(&ArgStr[1], &NewSave->SaveExpr);
    SetListLineVal(&NewSave->SaveExpr);
  }
  NewSave->Next = FirstIfSave;
  FirstIfSave = NewSave;
}


static void CodeCASE(void)
{
  Boolean eq;
  int z;

  if (!FirstIfSave) WrError(ErrNum_MissingIf);
  else if (ChkArgCnt(1, ArgCntMax))
  {
    if ((FirstIfSave->State != IfState_CASESWITCH) && (FirstIfSave->State != IfState_CASECASE)) WrError(ErrNum_InvIfConst);
    else
    {
      if (!FirstIfSave->SaveIfAsm)
        eq = True;
      else if (FirstIfSave->CaseFound)
        eq = False;
      else
      {
        TempResult t;

        as_tempres_ini(&t);
        eq = False;
        z = 1;
        do
        {
          EvalIfExpression(&ArgStr[z], &t);
          eq = (FirstIfSave->SaveExpr.Typ == t.Typ);
          if (eq)
           switch (t.Typ)
           {
             case TempInt:
               eq = (t.Contents.Int == FirstIfSave->SaveExpr.Contents.Int);
               break;
             case TempFloat:
               eq = (t.Contents.Float == FirstIfSave->SaveExpr.Contents.Float);
               break;
             case TempString:
               eq = (as_nonz_dynstr_cmp(&t.Contents.str, &FirstIfSave->SaveExpr.Contents.str) == 0);
               break;
             default:
               eq = False;
               break;
           }
          z++;
        }
        while (!eq && (z <= ArgCnt));
        as_tempres_free(&t);
      }
      IfAsm = (FirstIfSave->SaveIfAsm && eq & !FirstIfSave->CaseFound);
      if (FirstIfSave->SaveIfAsm)
        AddBoolFlag(eq && !FirstIfSave->CaseFound);
      if (eq)
        FirstIfSave->CaseFound = True;
      FirstIfSave->State = IfState_CASECASE;
    }
  }

  ActiveIF = !FirstIfSave || FirstIfSave->SaveIfAsm;
}


static void CodeELSECASE(void)
{
  if (ChkArgCnt(0, 0))
  {
    if ((FirstIfSave->State != IfState_CASESWITCH) && (FirstIfSave->State != IfState_CASECASE)) WrError(ErrNum_InvIfConst);
    else
      IfAsm = (FirstIfSave->SaveIfAsm && (!FirstIfSave->CaseFound));
    if (FirstIfSave->SaveIfAsm)
      AddBoolFlag(!FirstIfSave->CaseFound);
    FirstIfSave->CaseFound = True;
    FirstIfSave->State = IfState_CASEELSE;
  }

  ActiveIF = (!FirstIfSave) || (FirstIfSave->SaveIfAsm);
}


static void CodeENDCASE(void)
{
  if (!ChkArgCnt(0, 0));
  else if (!FirstIfSave) WrError(ErrNum_MissingIf);
  else
  {
    if ((FirstIfSave->State != IfState_CASESWITCH)
    && (FirstIfSave->State != IfState_CASECASE)
    && (FirstIfSave->State != IfState_CASEELSE)) WrError(ErrNum_InvIfConst);
    else
    {
      PIfSave NewSave;

      IfAsm = FirstIfSave->SaveIfAsm;
      if (!FirstIfSave->CaseFound) WrError(ErrNum_NoCaseHit);
      NewSave = FirstIfSave;
      FirstIfSave = NewSave->Next;
      as_snprintf(ListLine, STRINGSIZE, "[%u]", (unsigned)NewSave->StartLine);
      ifsave_free(NewSave);
    }
  }

  ActiveIF = IfAsm;
}


Boolean CodeIFs(void)
{
  Boolean Result = True;

  ActiveIF = False;

  switch (as_toupper(*OpPart.str.p_str))
  {
    case 'I':
      if (Memo("IF")) CodeIF();
      else if (Memo("IFDEF")) CodeIFDEF(False);
      else if (Memo("IFNDEF")) CodeIFDEF(True);
      else if (Memo("IFUSED")) CodeIFUSED(False);
      else if (Memo("IFNUSED")) CodeIFUSED(True);
      else if (Memo("IFEXIST")) CodeIFEXIST(False);
      else if (Memo("IFNEXIST")) CodeIFEXIST(True);
      else if (Memo("IFB")) CodeIFB(False);
      else if (Memo("IFNB")) CodeIFB(True);
      else Result = False;
      break;
    case 'E':
      if ((Memo("ELSE")) || (Memo("ELSEIF"))) CodeELSEIF();
      else if (Memo("ENDIF")) CodeENDIF();
      else if (Memo("ELSECASE")) CodeELSECASE();
      else if (Memo("ENDCASE")) CodeENDCASE();
      else Result = False;
      break;
    case 'S':
      if (Memo("SWITCH") && !SwitchIsOccupied) CodeSWITCH();
      else if (Memo("SELECT") && SwitchIsOccupied) CodeSWITCH();
      else Result = False;
      break;
    case 'C':
      if (Memo("CASE")) CodeCASE();
      else Result = False;
      break;
    default:
      Result = False;
  }

  return Result;
}

Integer SaveIFs(void)
{
  return (!FirstIfSave) ? 0 : FirstIfSave->NestLevel;
}

void RestoreIFs(Integer Level)
{
  PIfSave OldSave;

  while (FirstIfSave && (FirstIfSave->NestLevel != Level))
  {
    OldSave = FirstIfSave;
    FirstIfSave = OldSave->Next;
    IfAsm = OldSave->SaveIfAsm;
    ifsave_free(OldSave);
  }
}

Boolean IFListMask(void)
{
  switch (ListOn)
  {
    case 0:
      return True;
    case 1:
      return False;
    case 2:
      return ((!ActiveIF) && (!IfAsm));
    case 3:
      return (ActiveIF || (!IfAsm));
  }
  return True;
}

void AsmIFInit(void)
{
  IfAsm = True;
}

void asmif_init(void)
{
}

File: asmif.pas
{$I STDINC.PAS}
{$IFNDEF MSDOS}
{$C Moveable PreLoad Permanent }
{$ENDIF}
	UNIT AsmIF;

INTERFACE

	Uses
             {$IFDEF WINDOWS}
             WinDOS,Strings,
             {$ELSE}
             DOS,
             {$ENDIF}
             DecodeCm,
	     AsmDef,AsmSub,AsmPars;

TYPE
   PIfSave=^TIfSave;
   TIfSave=RECORD
	    Next:PIfSave;
            NestLevel:Integer;
	    SaveIfAsm:Boolean;
	    SaveExpr:TempResult;
	    State:(IfState_IFIF,IfState_IFELSE,
		   IfState_CASESWITCH,IfState_CASECASE,IfState_CASEELSE);
	    CaseFound:Boolean;
	   END;

VAR
   FirstIfSave:PIfSave;
   IfAsm:Boolean;         { FALSE: in einer neg. IF-Sequenz-->kein Code }


	FUNCTION CodeIFs:Boolean;

        FUNCTION SaveIFs:Integer;

        PROCEDURE RestoreIFs(Level:Integer);

        FUNCTION IFListMask:Boolean;

	PROCEDURE AsmIFInit;

IMPLEMENTATION

VAR
   ActiveIF:Boolean;    { ausgewertetes IF-Statement }

       FUNCTION GetIfVal(Cond:String):LongInt;
VAR
   IfOK:Boolean;
BEGIN
   FirstPassUnknown:=False;
   GetIfVal:=EvalIntExpression(Cond,Int32,IfOK);
   IF (FirstPassUnknown) OR (NOT IfOK) THEN
    BEGIN
     GetIfVal:=1;
     IF FirstPassUnknown THEN WrError(1820)
     ELSE IF NOT IfOK THEN WrError(1135);
    END;
END;

	PROCEDURE AddBoolFlag(Flag:Boolean);
BEGIN
   IF Flag THEN ListLine:='=>TRUE' ELSE ListLine:='=>FALSE';
END;

        PROCEDURE PushIF(IfExpr:LongInt);
VAR
   NewSave:PIfSave;
BEGIN
   New(NewSave);
   WITH NewSave^ DO
    BEGIN
     Next:=FirstIfSave;
     NestLevel:=SaveIFs+1;
     SaveIfAsm:=IfAsm;
     State:=IfState_IFIF; CaseFound:=IfExpr<>0;
    END;
   FirstIfSave:=NewSave;
   IfAsm:=IfAsm AND (IfExpr<>0);
END;

	PROCEDURE CodeIF;
VAR
   IfExpr:LongInt;
BEGIN
   ActiveIF:=IfAsm;

   IF NOT IfAsm THEN IfExpr:=1
   ELSE IF ArgCnt<>1 THEN
    BEGIN
     WrError(1110); IfExpr:=1;
    END
   ELSE IfExpr:=GetIfVal(ArgStr[1]);
   IF IfAsm THEN AddBoolFlag(IfExpr<>0);
   PushIF(IfExpr);
END;

	PROCEDURE CodeIFDEF;
VAR
   IfExpr:LongInt;
   Defined:Boolean;
BEGIN
   ActiveIF:=IfAsm;

   IF NOT IfAsm THEN IfExpr:=1
   ELSE IF ArgCnt<>1 THEN
    BEGIN
     WrError(1110); IfExpr:=1;
    END
   ELSE
    BEGIN
     Defined:=IsSymbolDefined(ArgStr[1]);
     IF IfAsm THEN
      IF Defined THEN ListLine:='=>DEFINED' ELSE ListLine:='=>UNDEFINED';
     IfExpr:=Ord(Defined XOR Memo('IFNDEF'));
    END;
   PushIF(IfExpr);
END;

	PROCEDURE CodeIFUSED;
VAR
   IfExpr:LongInt;
   Used:Boolean;
BEGIN
   ActiveIF:=IfAsm;

   IF NOT IfAsm THEN IfExpr:=1
   ELSE IF ArgCnt<>1 THEN
    BEGIN
     WrError(1110); IfExpr:=1;
    END
   ELSE
    BEGIN
     Used:=IsSymbolUsed(ArgStr[1]);
     IF Used THEN ListLine:='=>USED' ELSE ListLine:='=>UNUSED';
     IfExpr:=Ord(Used XOR Memo('IFNUSED'));
    END;
   PushIF(IfExpr);
END;

	PROCEDURE CodeIFEXIST;
VAR
   IfExpr:LongInt;
   Found:Boolean;
BEGIN
   ActiveIF:=IfAsm;

   IF NOT IfAsm THEN IfExpr:=1
   ELSE IF ArgCnt<>1 THEN
    BEGIN
     WrError(1110); IfExpr:=1;
    END
   ELSE
    BEGIN
     ArgPart:=ArgStr[1];
     IF ArgPart[1]='"' THEN Delete(ArgPart,1,1);
     IF ArgPart[Length(ArgPart)]='"' THEN Dec(Byte(ArgPart[0]));
     AddSuffix(ArgPart,IncSuffix);
     Found:=FSearch(ArgPart,'.;'+IncludeList)<>'';
     IF Found THEN ListLine:='=>FOUND' ELSE ListLine:='=>NOT FOUND';
     IfExpr:=Ord(Found XOR Memo('IFNEXIST'));
    END;
   PushIF(IfExpr);
END;

        PROCEDURE CodeIFB;
VAR
   IfExpr:LongInt;
   z:Integer;
   Blank:Boolean;
BEGIN
   ActiveIF:=IfAsm;

   IF NOT IfAsm THEN IfExpr:=1
   ELSE
    BEGIN
     Blank:=True;
     FOR z:=1 TO ArgCnt DO
      IF ArgStr[z]<>'' THEN Blank:=False;
     IF Blank THEN ListLine:='=>BLANK' ELSE ListLine:='NOT BLANK';
     IfExpr:=Ord(Blank XOR Memo('IFNB'));
    END;
   PushIF(IfExpr);
END;

	PROCEDURE CodeELSEIF;
VAR
   IfExpr:LongInt;
BEGIN
   IF FirstIfSave=Nil THEN WrError(1840)
   ELSE WITH FirstIfSave^ DO
    IF ArgCnt=0 THEN
     BEGIN
      IF State<>IfState_IFIF THEN WrError(1480)
      ELSE IF SaveIfAsm THEN
       BEGIN
	IfAsm:=NOT CaseFound;
	AddBoolFlag(NOT CaseFound);
       END;
      State:=IfState_IFELSE;
     END
    ELSE IF ArgCnt=1 THEN
     BEGIN
      IF State<>IfState_IFIF THEN WrError(1480)
      ELSE
       BEGIN
	IF NOT SaveIfAsm THEN IfExpr:=1
	ELSE IF CaseFound THEN IfExpr:=0
	ELSE IfExpr:=GetIfVal(ArgStr[1]);
	IfAsm:=SaveIfAsm AND (IfExpr<>0) AND (NOT CaseFound);
	IF SaveIfAsm THEN AddBoolFlag(IfExpr<>0);
	IF IfExpr<>0 THEN CaseFound:=True;
       END;
     END
    ELSE WrError(1110);

   ActiveIF:=(FirstIFSave=Nil) OR (FirstIfSave^.SaveIfAsm);
END;

	PROCEDURE CodeENDIF;
VAR
   NewSave:PIfSave;
BEGIN
   IF ArgCnt<>0 THEN WrError(1110);
   IF FirstIfSave=Nil THEN WrError(1840)
   ELSE WITH FirstIfSave^ DO
    BEGIN
     IF (State<>IfState_IFIF) AND (State<>IfState_IFELSE) THEN WrError(1480)
     ELSE
      BEGIN
       IfAsm:=SaveIfAsm;
       NewSave:=FirstIfSave; FirstIfSave:=NewSave^.Next;
       Dispose(NewSave);
      END
    END;

   ActiveIF:=IfAsm;
END;

	PROCEDURE EvalIfExpression(Cond:String; VAR erg:TempResult);
BEGIN
   FirstPassUnknown:=False;
   EvalExpression(Cond,erg);
   IF (erg.Typ=TempNone) OR (FirstPassUnknown) THEN
    BEGIN
     erg.Typ:=TempInt; Erg.Int:=1;
     IF FirstPassUnknown THEN WrError(1820);
    END;
END;

	PROCEDURE CodeSWITCH;
VAR
   NewSave:PIfSave;
BEGIN
   ActiveIF:=IfAsm;

   New(NewSave);
   WITH NewSave^ DO
    BEGIN
     Next:=FirstIfSave;
     NestLevel:=SaveIFs+1;
     SaveIfAsm:=IfAsm;
     CaseFound:=False; State:=IfState_CASESWITCH;
     IF ArgCnt<>1 THEN
      BEGIN
       SaveExpr.Typ:=TempInt; SaveExpr.Int:=1; IF (IfAsm) THEN WrError(1110);
      END
     ELSE
      BEGIN
       EvalIfExpression(ArgStr[1],SaveExpr);
       SetListLineVal(SaveExpr);
      END;
     FirstIfSave:=NewSave;
    END;
END;

	PROCEDURE CodeCASE;
VAR
   eq:Boolean;
   z:Byte;
   t:TempResult;
BEGIN
   IF FirstIfSave=Nil THEN WrError(1840)
   ELSE IF ArgCnt=0 THEN WrError(1110)
   ELSE WITH FirstIfSave^ DO
    BEGIN
     IF (State<>IfState_CASESWITCH) AND (State<>IfState_CASECASE) THEN WrError(1480)
     ELSE
      BEGIN
       IF NOT SaveIfAsm THEN eq:=True
       ELSE IF CaseFound THEN eq:=False
       ELSE
	BEGIN
	 eq:=False; z:=1;
	 REPEAT
	  EvalIfExpression(ArgStr[z],t);
	  eq:=(SaveExpr.Typ=t.Typ);
	  IF eq THEN
	   CASE t.Typ OF
	   TempInt:eq:=t.Int=SaveExpr.Int;
	   TempFloat:eq:=t.Float=SaveExpr.Float;
	   TempString:eq:=t.Ascii=SaveExpr.Ascii;
	   END;
	  Inc(z);
	 UNTIL (eq) OR (z>ArgCnt);
	END;
       IfAsm:=SaveIfAsm AND eq AND (NOT CaseFound);
       IF SaveIfAsm THEN AddBoolFlag(eq AND (NOT CaseFound));
       IF eq THEN CaseFound:=True;
       State:=IfState_CASECASE;
      END;
    END;

   ActiveIF:=(FirstIFSave=Nil) OR (FirstIFSave^.SaveIfAsm);
END;

	PROCEDURE CodeELSECASE;
BEGIN
   IF ArgCnt<>0 THEN WrError(1110)
   ELSE WITH FirstIfSave^ DO
    BEGIN
     IF (State<>IfState_CASESWITCH) AND (State<>IfState_CASECASE) THEN WrError(1480)
     ELSE IfAsm:=SaveIfAsm AND (NOT CaseFound);
     IF SaveIfAsm THEN AddBoolFlag(NOT CaseFound);
     CaseFound:=True;
     State:=IfState_CASEELSE;
    END;

   ActiveIF:=(FirstIFSave=Nil) OR (FirstIFSave^.SaveIfAsm);
END;

	PROCEDURE CodeENDCASE;
VAR
   NewSave:PIfSave;
BEGIN
   IF ArgCnt<>0 THEN WrError(1110);
   IF FirstIfSave=Nil THEN WrError(1840)
   ELSE WITH FirstIfSave^ DO
    BEGIN
     IF (State<>IfState_CASESWITCH) AND (State<>IfState_CASECASE) AND
	(State<>IfState_CASEELSE) THEN WrError(1480)
     ELSE
      BEGIN
       IfAsm:=SaveIfAsm;
       IF NOT CaseFound THEN WrError(100);
       NewSave:=FirstIfSave; FirstIfSave:=NewSave^.Next;
       Dispose(NewSave);
      END;
    END;

   ActiveIF:=IfAsm;
END;

	FUNCTION CodeIFs:Boolean;
BEGIN
   CodeIFs:=True; ActiveIF:=False;
   IF Memo('IF') THEN CodeIF
   ELSE IF (Memo('IFDEF')) OR (Memo('IFNDEF')) THEN CodeIFDEF
   ELSE IF (Memo('IFUSED')) OR (Memo('IFNUSED')) THEN CodeIFUSED
   ELSE IF (Memo('IFEXIST')) OR (Memo('IFNEXIST')) THEN CodeIFEXIST
   ELSE IF (Memo('IFB')) OR (Memo('IFNB')) THEN CodeIFB
   ELSE IF Memo('ELSEIF') THEN CodeELSEIF
   ELSE IF Memo('ENDIF') THEN CodeENDIF
   ELSE IF Memo('SWITCH') THEN CodeSWITCH
   ELSE IF Memo('CASE') THEN CodeCASE
   ELSE IF Memo('ELSECASE') THEN CodeELSECASE
   ELSE IF Memo('ENDCASE') THEN CodeENDCASE
   ELSE CodeIFs:=False;
END;

        FUNCTION SaveIFs:Integer;
BEGIN
   IF FirstIfSave=Nil THEN SaveIFs:=0
   ELSE SaveIFs:=FirstIfSave^.NestLevel;
END;

        PROCEDURE RestoreIFs(Level:Integer);
VAR
   OldSave:PIfSave;
BEGIN
   WHILE (FirstIfSave<>Nil) AND (FirstIfSave^.NestLevel<>Level) DO
    BEGIN
     OldSave:=FirstIfSave; FirstIfSave:=OldSave^.Next;
     IfAsm:=OldSave^.SaveIfAsm;
     Dispose(OldSave);
    END;
END;

        FUNCTION IFListMask:Boolean;
BEGIN
   CASE ListOn OF
   0:IFListMask:=True;
   1:IFListMask:=False;
   2:IFListMask:=(NOT ActiveIF) AND (NOT IfAsm);
   3:IFListMask:=ActiveIF OR (NOT IfAsm);
   END;
END;

        PROCEDURE AsmIFInit;
BEGIN
   IfAsm:=True;
END;

END.
